enhancements/machine-config/machine-config-node.md

```diff
@@ -0,0 +1,340 @@
+ip-10-0-51-162.ec2.internal   True      False             False              False              False              False      False                     False                False          False           False           False
+ip-10-0-59-150.ec2.internal   False     False             False              True               False              False      False                     False                False          False           True            False
+```
+
+Where each name represents a node in a given machineconfigpool. The statuses reported are created explicitly from MCO node annotations and MCO actions, no other operator actions are taken into account here. This allows us to get quite specific in what is occuring on the nodes.
+
+```console
+Name:         ip-10-0-12-194.ec2.internal
+Namespace:    
+Labels:       <none>
+Annotations:  <none>
+API Version:  machineconfiguration.openshift.io/v1alpha1
+Kind:         MachineConfigNode
+Metadata:
+  Creation Timestamp:  2023-10-17T13:08:58Z
+  Generation:          1
+  Resource Version:    49443
+  UID:                 4bd758ab-2187-413c-ac42-882e61761b1d
+Spec:
+  Node Ref:
+    API Version:  
+    Kind:         Node
+    Name:         ip-10-0-12-194.ec2.internal
+  Pool:
+    API Version:  machineconfiguration.openshift.io/v1
+    Kind:         MachineConfigPool
+    Name:         master
+Status:
+  Conditions:
+    Last Transition Time:  2023-10-17T13:09:02Z
+    Message:               Node has completed update to config rendered-master-cf99e619747ab19165f11e3546c71f1e
+    Reason:                NodeUpgradeComplete
+    Status:                True
+    Type:                  Updated
+    Last Transition Time:  2023-10-17T13:09:02Z
+    Message:               This node has not yet entered the UpdatePreparing phase
+    Reason:                NotYetOccured
+    Status:                False
+  Config Version:
+    Current:            rendered-worker-823ff8dc2b33bf444709ed7cd2b9855b
+    Desired:            rendered-worker-823ff8dc2b33bf444709ed7cd2b9855b
+  Health:               Healthy
+  Most Recent Error:    
+  Observed Generation:  3
+```
+
+The above struct gives us some helpful information about a node as it pertains to the MCO. First, all upgrade related events that have occured on the node from the most recent uprade process (no matter how small) appear in the `conditions`. Other than that, you can see the current and desired machineconfig indicating whether or not the node should be updating as well as whether or not the currently tracked update process held in `conditions` is updating to the expected MachineConfig. The Node refernece and ObservedGeneration exist to let the user know some extra information on the object and how many times we have gone throgh some upgrade related changes.
```

Comment:
nit: "refernece" -> "reference".

> The Node refernece and ObservedGeneration exist to let the user know some extra information on the object and how many times we have gone throgh some upgrade related changes.

I'd expect the reasons are:

* The Node reference allows you to figure out which node the MachineConfigNode is about, if the relationship isn't "we always name MachineConfigNode to match the node name".  But if the names always match, I'm not sure what the `spec` node reference would deliver.  And I'm not clear on why it's part of the `spec`.  It seems like this structure could have no `spec` knobs at all, and just be something that the controller that creates it uses to share `status` information.
* ObservedGeneration is often a way to decide if the `status` on the resource was written after consuming the most recent `metadata`/`spec` bumps.  But in this case, we don't expect anyone to be bumping the resource except for the `status`-writing controller.  So maybe we don't need this property?  Or maybe we want to use it, or something with a similar name, to record the `generation` we say last on the Node, when we last scraped that resource for feedback set by the machine-config daemon.  That would allow us to say "the MCD and other actors keep bumping the Node, but the MCN controller doesn't seem to be keeping up with those changes, and I suspect some of the information in this MCN `status` may be stale as a result".

Reply:
This field is intended to be updated when the desiredConfig in the spec gets bumped to a new value. This, of course, indicates that the status desiredConfig will change as well.

This happens when the MCO (specifically the MCD) triggers an upgrade and updates the node annotations that the MCO owns to indicate that the desired and currentconfig differ.

I think this is helpful for indicating how many upgrade cycles we have gone through

Reply:
I think @cdoern and I spoke offline after this comment was made, there's a flow that we should be accounting for where the spec desired config gets updated, then the MCD checks that there is a valid transition between the spec desired and the current, and only then updates the status desired.

This gives a signal to the user that the spec change has been validated and accepted and is otherwise not represented. The observedGeneration should be updated once the new spec.desired has been observed (as it's the only mutable field in spec).

If there's an error or not, only once the "validation" has been done should the observed generation be updated, along with either a happy or sad status based on the spec change

Reply:
the current implementation does this, so I will add that in and describe how it is done.